EVENTS.md

```diff
@@ -0,0 +1,416 @@
+undertake, we also perform a number of further validation checks on the contracts:
+
+* No attribute key may start with `_`. This is currently used for `_contract_address` and `_code_id` and is reserved for a 
+  namespace for injecting more *trusted* attributes from the `x/wasm` module as opposed to the contract itself
+* Event types are trimmed of whitespace, and must have at least two characters prior to prepending `wasm-`. If the contract returns
+  "  hello\n", the event type will look like `wasm-hello`. If it emits "  a  ", this will be rejected with an error (aborting execution!)
+* Attribute keys and values (both in `attributes` and under `events`) are trimmed of leading/trailing whitespace. If they are empty after
+  trimming, they are rejected as above (aborting the execution). Otherwise, they are passed verbatim.
+
+## Event Details for wasmd
+
+Beyond the basic Event system and emitted events, we must handle more advanced cases in `x/wasm`
+and thus add some more logic to the event processing. Remember that CosmWasm contracts dispatch other
+messages themselves, so far from the flattened event structure, or even a list of list (separated
+by message index in the tx), we actually have a tree of messages, each with their own events. And we must
+flatten that in a meaningful way to feed it into the event system.
+
+Furthermore, with the sub-message reply handlers, we end up with eg. "Contract A execute", "Contract B execute",
+"Contract A reply". If we return all events by all of these, we may end up with many repeated event types and
+a confusing results. However, we may use the standard "message" events to separate the sub-messages as it marks
+where the next one starts. With careful analysis of the "sender" field on these "message" markers, we may be able
+to reconstruct much of the tree execution path. We should ensure all this information is exposed in the most
+consistent way possible.
+
+### Combining Events from Sub-Messages
+
+Each time a contract is executed, it not only returns the `message` event from its call, the `execute` event for the
+contact and the `wasm` event with any custom fields from the contract itself. It will also return the same set of information
+for all messages that it returned, which were later dispatched.
+
+Until 0.18.0, we stripped out the `message` events that were returned when dispatching `Response.messages`. However,
+they contain useful information to trace the topography of the call, especially when we call into native modules,
+and it makese sense to maintain the full information.
+
+Note that these are all appended in order called. So if I execute a contract, which returns two messages, one to instantiate a new
+contract and the other to set the withdrawl address, while also using `ReplyOnSuccess` for the instantiation (to get the
+address), it will emit a series of events that looks something like this:
+
+```go
+/// original execution (top-level message)
```

Comment:
This list of events describe the process flow but as events are emitted **after** execution the actual order would be different. I think L321 would be the last event emitted when the original message was processed. L330 would be the first one when we handle the first contract response.

Reply:
We can change what happens. I want to document the desired behaviour in here (which will be actual behaviour with the 0.18.0 tag), not the current behaviour.

We have control in x/wasm to do something like:

* Keeper.execute called
* call wasmvm.execute
* emit `execute` message including the data from the response
* emit `wasm` and `wasm-*` events from the contract
* for each message in Response:
  * emit some "start sub message" event (if desired)
  * call the execution logic here (and emit all events like this, recursively - filtering message event)
  * emit some "stop sub message event"?
* emit any other events we wish

Reply:
As to the `message: module=wasm`, this is emitted in the message server and we can perform this any way we want to.
We can add it top the event manager before calling the keeper, or afterwards. I think placing it first (as in distribution) makes more sense than at the end (in bank).

What do you think the *desired* behaviour here is?

Reply:
IMHO to emit the events in order of processing is better than what we do now. It should not matter but for debugging and flattening, I can see the benefit.